In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [2]:
df= pd.read_csv('Table2_3_model_floats.csv')

In [3]:
import sweetviz as sv
report = sv.analyze(df)
report.show_html('sweetviz_report_all.html')

                                             |      | [  0%]   00:00 -> (? left)

Report sweetviz_report_all.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


In [6]:
df.isnull().sum().sort_values()

Unnamed: 0                               0
RSE licensee board structure             0
RSE licensee profit status               0
Fund's RSE licensee                      0
RSE Membership base                      0
                                      ... 
Growth in number of member accounts    356
Total number of member accounts        447
Ten-year\nrate of return               455
Total members' benefits                462
Average member account balance         473
Length: 82, dtype: int64

In [9]:
df['Period'] = pd.to_datetime(df['Period'], format='%Y-%m-%d')


In [10]:
df['Year'] = df['Period'].dt.year
df.head()

Unnamed: 0     Period                              Fund name          ABN  \
0           0 2022-06-30     Alcoa of Australia Retirement Plan  80928800255   
1           1 2022-06-30                              AMG Super  30099320583   
2           2 2022-06-30                         AMP Super Fund  78421957449   
3           3 2022-06-30             Aracon Superannuation Fund  40586548205   
4           4 2022-06-30  ASGARD Independence Plan Division Two  90194410365   

  RSE Regulatory classification  Fund type RSE Membership base  \
0              Non public offer  Corporate      Corporate base   
1                  Public offer     Retail        General base   
2                  Public offer     Retail        General base   
3                  Public offer     Retail        General base   
4                  Public offer     Retail        General base   

                          Fund's RSE licensee  \
0  Alcoa of Australia Retirement Plan Pty Ltd   
1      Equity Trustees Superannuation Limited   
2    N. M. Superannuation Proprietary Limited   
3      Equity Trustees Superannuation Limited   
4                 BT Funds Management Limited   

                      RSE licensee ownership type RSE licensee profit status  \
0  Employer sponsor (non-public sector) ownership      Not for profit status   
1        Financial services corporation ownership          For profit status   
2        Financial services corporation ownership          For profit status   
3        Financial services corporation ownership          For profit status   
4        Financial services corporation ownership          For profit status   

   ... Advice expenses  of which: associated with intrafund advice  \
0  ...             0.0                                         0.0   
1  ...          6256.0                                         0.0   
2  ...         36761.0                                         0.0   
3  ...          2595.0                                         0.0   
4  ...             0.0                                         0.0   

   Net earnings   Income tax expense / benefit  Net earnings after tax  \
0       -53078.0                       -8602.0                -44476.0   
1      -138740.0                      -23474.0               -115266.0   
2     -3653121.0                     -396125.0              -3256996.0   
3       -43557.0                       -3504.0                -40053.0   
4     -1339999.0                     -203957.0              -1136042.0   

   Cash flow adjusted net assets  Total assets at the end of period  \
0                      2258997.0                          2172137.0   
1                      1607575.0                          1553647.0   
2                     60581015.0                         55838175.0   
3                       586242.0                           595099.0   
4                     20424719.0                         16839656.0   

   Total liabilities at the end of period  Net assets at the end of period  \
0                                 20071.0                        2152066.0   
1                                  5396.0                        1548251.0   
2                                964296.0                       54873879.0   
3                                  3414.0                         591685.0   
4                                 41287.0                       16798369.0   

   Year  
0  2022  
1  2022  
2  2022  
3  2022  
4  2022  

[5 rows x 83 columns]

In [12]:
df.dropna(inplace=True)

In [23]:
df.columns

Index(['Unnamed: 0', 'Period', 'Fund name', 'ABN',
       'RSE Regulatory classification', 'Fund type', 'RSE Membership base',
       'Fund's RSE licensee', 'RSE licensee ownership type',
       'RSE licensee profit status', 'RSE licensee board structure',
       'Total assets', 'Number of investment options',
       'One-year\nrate of return_x', 'Five-year\nrate of return',
       'Ten-year\nrate of return', 'Total number of member accounts',
       'Total members' benefits', 'Average member account balance',
       'Growth in number of member accounts',
       'Net assets at beginning of period', 'Employer contributions ',
       'Personal contributions', 'Government co-contributions',
       'Low income super contributions', 'Other member contributions',
       'Member contributions', 'Defined benefit contributions',
       'Contribution tax and surcharge', 'Total Members' benefit flows in',
       'Inward rollovers', 'Outward rollover', 'Net rollovers',
       'Lump sum benefit pay

In [36]:
feature_cols = ["Year", 'Total number of member accounts',
    "Operating expense ratio_y",
    "Employer contributions ",
    "Total investment income",'Number of investment options',
                     'Net earnings after tax','Net insurance flows', 'Member contributions'
]


In [37]:
# Define a function that accepts a list of features in a DataFrame and returns the RMSE for a test dataset.

def train_test_rmse(df, feature_cols):
    X = df[feature_cols]
    y = df['Ten-year\nrate of return']
    
    X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2,random_state=123)
    
    linreg = LinearRegression()
    linreg.fit(X_train, y_train)
    
    y_pred = linreg.predict(X_test)
    return np.sqrt(metrics.mean_squared_error(y_test, y_pred))


In [42]:
from sklearn import metrics

# Compare different sets of features.
print(train_test_rmse(df, ["Year", 
                           'Total number of member accounts',
                           "Operating expense ratio_y",
                           "Employer contributions ",
                           "Total investment income",'Number of investment options',
                           'Net earnings after tax',
                           'Net insurance flows', 
                           'Member contributions',
                           "Total investment income"]))
print(train_test_rmse(df, ["Year",'Total number of member accounts',
                           "Operating expense ratio_y",
                           "Employer contributions ",
                           "Total investment income",
                           'Net earnings after tax',
                           'Net insurance flows', 
                           'Member contributions',
                           "Total investment income"]))
print(train_test_rmse(df, ["Year",
                           "Operating expense ratio_y",
                           "Employer contributions ",
                           "Total investment income",
                           
                           'Net insurance flows', 
                           'Member contributions',
                           "Total investment income"]))

0.012473917814536202
0.012182625293940822
0.012303065132438087


In [43]:
#get dummies 
dummy_df = pd.get_dummies(df['Fund type'], prefix='Fund_type')

# Concatenate the dummy variables with the original DataFrame
df = pd.concat([df, dummy_df], axis=1)

df.drop('Fund type', axis=1, inplace=True)

In [44]:
dummy_year = pd.get_dummies(df['Year'], prefix='Year')

# Concatenate the dummy variables with the original DataFrame
df = pd.concat([df, dummy_year], axis=1)

df.drop('Year', axis=1, inplace=True)

In [45]:
dummy_structure = pd.get_dummies(df['RSE licensee board structure'], prefix='Board_Structure')

# Concatenate the dummy variables with the original DataFrame
df = pd.concat([df, dummy_structure], axis=1)

df.drop('RSE licensee board structure', axis=1, inplace=True)

In [46]:
df.columns


Index(['Unnamed: 0', 'Period', 'Fund name', 'ABN',
       'RSE Regulatory classification', 'RSE Membership base',
       'Fund's RSE licensee', 'RSE licensee ownership type',
       'RSE licensee profit status', 'Total assets',
       'Number of investment options', 'One-year\nrate of return_x',
       'Five-year\nrate of return', 'Ten-year\nrate of return',
       'Total number of member accounts', 'Total members' benefits',
       'Average member account balance', 'Growth in number of member accounts',
       'Net assets at beginning of period', 'Employer contributions ',
       'Personal contributions', 'Government co-contributions',
       'Low income super contributions', 'Other member contributions',
       'Member contributions', 'Defined benefit contributions',
       'Contribution tax and surcharge', 'Total Members' benefit flows in',
       'Inward rollovers', 'Outward rollover', 'Net rollovers',
       'Lump sum benefit payments ', 'Pension benefit payments',
       'Total b

In [49]:
print(train_test_rmse(df, ['Total number of member accounts',
                           "Operating expense ratio_y",
                           "Employer contributions ",
                           "Total investment income",
                           'Net earnings after tax',
                           'Net insurance flows', 
                           'Member contributions',
                           "Total investment income",'Fund_type_Corporate',
       'Fund_type_Industry', 'Fund_type_Public Sector', 'Fund_type_Retail',
       'Fund_type_Retail - ERF', 'Year_2015', 'Year_2016', 'Year_2017',
       'Year_2018', 'Year_2019', 'Year_2020', 'Year_2021', 'Year_2022',
       'Board_Structure_Equal representation required by governing rules',
       'Board_Structure_Equal representation required by legislation',
       'Board_Structure_Non equal representation']))

0.009285865730768367


In [50]:
print(train_test_rmse(df, ['Total number of member accounts',
                           "Operating expense ratio_y",
                           "Employer contributions ",
                           "Total investment income",
                           'Net earnings after tax',
                           'Net insurance flows', 
                           'Member contributions',
                           "Total investment income",'Fund_type_Corporate',
       'Fund_type_Industry', 'Fund_type_Public Sector', 'Fund_type_Retail',
       'Fund_type_Retail - ERF', 'Year_2015', 'Year_2016', 'Year_2017',
       'Year_2018', 'Year_2019', 'Year_2020', 'Year_2021', 'Year_2022'
      ]))

0.009284211883761313


In [52]:
#null RMSE as a baseline model 
feature_cols = ['Total number of member accounts',
                           "Operating expense ratio_y",
                           "Employer contributions ",
                           "Total investment income",
                           'Net earnings after tax',
                           'Net insurance flows', 
                           'Member contributions',
                           "Total investment income",'Fund_type_Corporate',
       'Fund_type_Industry', 'Fund_type_Public Sector', 'Fund_type_Retail',
       'Fund_type_Retail - ERF', 'Year_2015', 'Year_2016', 'Year_2017',
       'Year_2018', 'Year_2019', 'Year_2020', 'Year_2021', 'Year_2022']
X = df[feature_cols]
y = df['Ten-year\nrate of return']

X_train, X_test, y_train, y_test = train_test_split(X, y,random_state=123)

linreg = LinearRegression()
linreg.fit(X_train, y_train)
#y_test.mean() gives you the average number of bike rentals in your testing dataset

y_test.mean()


0.05708708222891566

In [53]:
# Create a NumPy array with the same shape as y_test.
y_null = np.zeros_like(y_test, dtype=float)
#This creates a new NumPy array with the same shape as y_test and
#sets all its elements to zero.


y_null.fill(y_test.mean())

# Calculate RMSE for the null model
rmse_null = np.sqrt(mean_squared_error(y_test, y_null))
rmse_null

0.015491328146092465

Okay, my model is doing better than predicting the mean value for y_test array. 

I learnt with previous model try, that to enhance the performance of my machine learning model, it's crucial to balance the impact of each feature. This involves scaling them so that no single feature unduly influences the model due to its large numerical range.If we do not to scale the features, the algorithm could give disproportionate weight to those with larger numerical values.

In [54]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression

# Initialize the StandardScaler
scaler = StandardScaler()


# You can now use X_train_scaled and X_test_scaled in your model training

feature_cols = ['Total number of member accounts',
                           "Operating expense ratio_y",
                           "Employer contributions ",
                           "Total investment income",
                           'Net earnings after tax',
                           'Net insurance flows', 
                           'Member contributions',
                           "Total investment income",'Fund_type_Corporate',
       'Fund_type_Industry', 'Fund_type_Public Sector', 'Fund_type_Retail',
       'Fund_type_Retail - ERF', 'Year_2015', 'Year_2016', 'Year_2017',
       'Year_2018', 'Year_2019', 'Year_2020', 'Year_2021', 'Year_2022']

X = df[feature_cols]
y = df['Ten-year\nrate of return']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

linreg = LinearRegression()
linreg.fit(X_train_scaled, y_train)

# Make predictions and evaluate the model
y_pred = linreg.predict(X_test_scaled)


# Get the intercept and coefficients
intercept = linreg.intercept_
coefficients = linreg.coef_

print(f"Intercept: {intercept}")
print("Coefficients:")
for feature, coef in zip(feature_cols, coefficients):
    print(f"{feature}: {coef}")

Intercept: 0.05394903450314466
Coefficients:
Total number of member accounts: -0.0004060153170993916
Operating expense ratio_y: -0.0065774013660163545
Employer contributions : 0.002554808630130066
Total investment income: 0.0002679489732784413
Net earnings after tax: 0.0006313326308123282
Net insurance flows: -0.0006387861396809381
Member contributions: -0.0021242269008096104
Total investment income: 0.0002679489732784426
Fund_type_Corporate: 0.0023193748615716546
Fund_type_Industry: 0.0024258435306719534
Fund_type_Public Sector: 0.0004956235643689708
Fund_type_Retail: -0.0026745446733879522
Fund_type_Retail - ERF: -0.002835983067395097
Year_2015: -0.0008752760430811884
Year_2016: -0.004388991195782826
Year_2017: -0.006519450556007079
Year_2018: -0.0010979337600597406
Year_2019: 0.005366511724993659
Year_2020: 0.0020319845872896975
Year_2021: 0.004384323946967617
Year_2022: 0.0029010960865402118


In [56]:
def calculate_rmse(y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    return rmse
rmse = calculate_rmse(y_test, y_pred)
print("RMSE for model with scaled features:", rmse)

RMSE for model with scaled features: 0.009284211886941418


In [57]:
r_squared = linreg.score(X_test_scaled, y_test)
print("R-squared value:", r_squared)

R-squared value: 0.6327012973826398


R-squared value of approximately 0.63 indicates that around 63.3% of the variance in my targe variable (10Y ROR) can be explained by the independent variables in my model. This suggests a moderately strong fit: my model captures a good portion of the information in the data, but there is still around 36.7% of the variance that is unaccounted for. 

In [58]:
#RANKING MODEL 

from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error
from math import sqrt

# Initialize the Gradient Boosting Regressor
gbm = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=123)

# Fit the model
gbm.fit(X_train_scaled, y_train)

# Make predictions on the test set
y_pred_gbm = gbm.predict(X_test_scaled)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred_gbm)
rmse = sqrt(mse)  # Root Mean Square Error

print(f"Mean Squared Error (MSE): {mse}")
print(f"Root Mean Square Error (RMSE): {rmse}")

Mean Squared Error (MSE): 6.001394156667522e-05
Root Mean Square Error (RMSE): 0.00774686656440365


GBR is doing better than previous model. Now it's time to predict the ranking of funds, using 10 ROR as the relevance score.

In [59]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split


# Separate the "Fund name" column before splitting
fund_names = df['Fund name']

X = df[feature_cols]
y = df['Ten-year\nrate of return']

# Split the data, including the fund_names to keep indices consistent
X_train, X_test, y_train, y_test, fund_names_train, fund_names_test = train_test_split(X, y, fund_names, test_size=0.2, random_state=123)

# Scale the features
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Initialize and train GradientBoostingRegressor
gbr = GradientBoostingRegressor()
gbr.fit(X_train_scaled, y_train)

# Make predictions
y_pred = gbr.predict(X_test_scaled)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"Root Mean Square Error (RMSE): {rmse}")

# Sort the predictions (in descending order for highest returns first)
sorted_indices = np.argsort(y_pred)[::-1]

# Sort fund names according to the sorted prediction indices
sorted_fund_names = fund_names_test.iloc[sorted_indices].reset_index(drop=True)

# Create a DataFrame to hold sorted fund names and their predicted 10-year returns
sorted_df = pd.DataFrame({
    'Fund Name': sorted_fund_names,
    'Predicted 10-Year Return': y_pred[sorted_indices]
})

print("Sorted Fund Names and Predicted 10-Year Returns:")
print(sorted_df)

Root Mean Square Error (RMSE): 0.0077536091830397854
Sorted Fund Names and Predicted 10-Year Returns:
                                             Fund Name  \
0    CONSTRUCTION AND BUILDING UNIONS SUPERANNUATIO...   
1                         Sunsuper Superannuation Fund   
2         Military Superannuation & Benefits Fund No 1   
3                                           Care Super   
4       Energy Industries Superannuation Scheme-Pool B   
..                                                 ...   
194                     SuperTrace Superannuation Fund   
195                  Macquarie ADF Superannuation Fund   
196                            Map Superannuation Plan   
197                     SuperTrace Superannuation Fund   
198         IOOF Portfolio Service Superannuation Fund   

     Predicted 10-Year Return  
0                    0.083731  
1                    0.083170  
2                    0.082688  
3                    0.080300  
4                    0.080113  
..       

Filtering the results only to show industry type funds:

In [63]:
# Create a DataFrame to combine all the sorted test data
sorted_df = pd.DataFrame({
    'Fund Name': sorted_fund_names,
    'Predicted 10y Return': sorted_predictions,
    'Fund_type_Industry': fund_type_industry_test.iloc[sorted_indices].reset_index(drop=True)
})

# Filter the DataFrame to show only rows where Fund_type_Industry == 1
filtered_df = sorted_df[sorted_df['Fund_type_Industry'] == 1]

print(filtered_df)


                                             Fund Name  Predicted 10y Return  \
29                             TWU Superannuation Fund              0.042243   
33                                          Care Super              0.043807   
35   The Victorian Independent Schools Superannuati...              0.044187   
37   Australian Catholic Superannuation and Retirem...              0.044507   
44                                            Unisuper              0.046320   
48                        Austsafe Superannuation Fund              0.046860   
51                        Sunsuper Superannuation Fund              0.047752   
52                                         Media Super              0.048021   
54                                         First Super              0.048109   
56   The Victorian Independent Schools Superannuati...              0.048666   
57                                 Combined Super Fund              0.048722   
59   Building Unions Superannuation Sche

In [65]:
filtered_df.sort_values('Predicted 10y Return', ascending=False)[['Fund Name', 'Predicted 10y Return']].head(10)


Fund Name  Predicted 10y Return
198  CONSTRUCTION AND BUILDING UNIONS SUPERANNUATIO...              0.085493
197                       Sunsuper Superannuation Fund              0.085287
192                                       Spirit Super              0.081111
189  Building Unions Superannuation Scheme (Queensl...              0.079010
187                                              HESTA              0.078259
186                                        Prime Super              0.078028
185                                          Rei Super              0.077745
184                       HOSTPLUS Superannuation Fund              0.077569
182                                 Intrust Super Fund              0.077030
180                                         Care Super              0.076841